File: android/pouchprov-core/src/main/kotlin/io/golioth/pouchprov/Flows.kt
// Copyright (c) 2026 Jonathan Beri
// SPDX-License-Identifier: Apache-2.0

package io.golioth.pouchprov

import io.golioth.pouchprov.session.ProvSession
import kotlinx.coroutines.delay

/**
 * High-level provisioning flows composed over a [ProvSession]. A faithful port
 * of the Python `flows` module.
 */
object Flows {

    suspend fun getVersion(session: ProvSession): VersionInfo {
        val info = Messages.decodeVerRsp(session.request(Messages.PATH_VER, Messages.encodeVerReq()))
        if (info.proto != Messages.PROTO_VERSION) {
            throw IllegalStateException("unsupported device protocol v${info.proto}")
        }
        session.blockSize = info.blockSize
        return info
    }

    suspend fun authorizeIfNeeded(session: ProvSession, info: VersionInfo, pop: String?) {
        if (info.popRequired) {
            if (pop.isNullOrEmpty()) throw IllegalStateException("device requires a proof-of-possession (pop)")
            Auth.authorize(session, pop)
        }
    }

    /** Trigger a scan, poll until finished, and collect all result pages. */
    suspend fun scan(session: ProvSession, timeoutMs: Long = 20_000): List<ScanEntry> {
        Messages.decodeScanStartRsp(session.request(Messages.PATH_SCAN, Messages.encodeScanStart()))

        val deadline = nowMs() + timeoutMs
        var total: Int
        while (true) {
            val (finished, t) = Messages.decodeScanStatusRsp(
                session.request(Messages.PATH_SCAN, Messages.encodeScanGetStatus()),
            )
            total = t
            if (finished) break
            if (nowMs() > deadline) throw IllegalStateException("scan did not finish")
            delay(500)
        }

        val results = ArrayList<ScanEntry>()
        while (results.size < total) {
            val page = Messages.decodeScanResultsRsp(
                session.request(Messages.PATH_SCAN, Messages.encodeScanGetResults(results.size, 6)),
            )
            if (page.isEmpty()) break
            results.addAll(page)
        }
        return results
    }

    /** Set + apply credentials, then poll status until connected or failed. */
    suspend fun configureWifi(
        session: ProvSession,
        ssid: ByteArray,
        password: ByteArray?,
        timeoutMs: Long = 40_000,
    ): WifiStatus {
        Messages.decodeConfigSetRsp(
            session.request(Messages.PATH_CONFIG, Messages.encodeConfigSet(ssid, password)),
        )
        Messages.decodeConfigApplyRsp(
            session.request(Messages.PATH_CONFIG, Messages.encodeConfigApply()),
        )

        val deadline = nowMs() + timeoutMs
        while (true) {
            val status = Messages.decodeConfigStatusRsp(
                session.request(Messages.PATH_CONFIG, Messages.encodeConfigGetStatus()),
            )
            if (status.state == StaState.CONNECTED || status.state == StaState.FAILED) return status
            if (nowMs() > deadline) throw IllegalStateException("connection did not settle")
            delay(1_000)
        }
    }

    /** Write one DER credential to the device in ordered chunks. */
    suspend fun pushCredential(session: ProvSession, kind: CredKind, der: ByteArray, chunk: Int = 256) {
        val total = der.size
        var off = 0
        while (off < total) {
            val piece = der.copyOfRange(off, minOf(off + chunk, total))
            val received = Messages.decodeCredWriteRsp(
                session.request(Messages.PATH_CRED, Messages.encodeCredWrite(kind, off, total, piece)),
            )
            off += piece.size
            if (received != off) throw IllegalStateException("cred write desync: device has $received, sent $off")
        }
    }

    /** Push device cert + key (+ optional CA), then finalize. */
    suspend fun bootstrapCredentials(
        session: ProvSession,
        certDer: ByteArray,
        keyDer: ByteArray,
        caDer: ByteArray? = null,
    ) {
        pushCredential(session, CredKind.DEVICE_CERT, certDer)
        pushCredential(session, CredKind.PRIVATE_KEY, keyDer)
        if (caDer != null) pushCredential(session, CredKind.CA_CERT, caDer)
        Messages.decodeCredFinalizeRsp(session.request(Messages.PATH_CRED, Messages.encodeCredFinalize()))
    }

    suspend fun endSession(session: ProvSession) {
        Messages.decodeCtrlRsp(
            session.request(Messages.PATH_CTRL, Messages.encodeCtrl(CtrlOp.END)),
            CtrlOp.END,
        )
    }

    private fun nowMs(): Long = System.nanoTime() / 1_000_000
}

File: ios/PouchProv/Sources/PouchProvCore/Flows.swift
// Copyright (c) 2026 Jonathan Beri
// SPDX-License-Identifier: Apache-2.0

import Foundation

/// A provisioning flow reached an unexpected state (the analogue of the
/// Kotlin `IllegalStateException` in `Flows.kt`).
public struct FlowError: Error, CustomStringConvertible {
    public let message: String

    public init(_ message: String) {
        self.message = message
    }

    public var description: String { message }
}

/// High-level provisioning flows composed over a ``ProvSession``. A faithful
/// port of the Kotlin/Python `Flows`.
public enum Flows {

    public static func getVersion(_ session: ProvSession) async throws -> VersionInfo {
        let info = try Messages.decodeVerRsp(
            try await session.request(path: Messages.pathVer, message: Messages.encodeVerReq())
        )
        guard info.proto == Messages.protoVersion else {
            throw FlowError("unsupported device protocol v\(info.proto)")
        }
        session.blockSize = info.blockSize
        return info
    }

    public static func authorizeIfNeeded(_ session: ProvSession, info: VersionInfo, pop: String?) async throws {
        if info.popRequired {
            guard let pop, !pop.isEmpty else {
                throw FlowError("device requires a proof-of-possession (pop)")
            }
            try await Auth.authorize(session: session, pop: pop)
        }
    }

    /// Trigger a scan, poll until finished, and collect all result pages.
    public static func scan(_ session: ProvSession, timeoutMs: Int64 = 20_000) async throws -> [ScanEntry] {
        try Messages.decodeScanStartRsp(
            try await session.request(path: Messages.pathScan, message: Messages.encodeScanStart())
        )

        let deadline = MonotonicClock.nowMs() + timeoutMs
        var total: Int
        while true {
            let (finished, t) = try Messages.decodeScanStatusRsp(
                try await session.request(path: Messages.pathScan, message: Messages.encodeScanGetStatus())
            )
            total = t
            if finished { break }
            guard MonotonicClock.nowMs() <= deadline else { throw FlowError("scan did not finish") }
            try await Task.sleep(nanoseconds: 500 * 1_000_000)
        }

        var results: [ScanEntry] = []
        while results.count < total {
            let page = try Messages.decodeScanResultsRsp(
                try await session.request(
                    path: Messages.pathScan,
                    message: Messages.encodeScanGetResults(start: results.count, count: 6)
                )
            )
            if page.isEmpty { break }
            results.append(contentsOf: page)
        }
        return results
    }

    /// Set + apply credentials, then poll status until connected or failed.
    public static func configureWifi(
        _ session: ProvSession,
        ssid: Data,
        password: Data?,
        timeoutMs: Int64 = 40_000
    ) async throws -> WifiStatus {
        try Messages.decodeConfigSetRsp(
            try await session.request(
                path: Messages.pathConfig,
                message: Messages.encodeConfigSet(ssid: ssid, password: password)
            )
        )
        try Messages.decodeConfigApplyRsp(
            try await session.request(path: Messages.pathConfig, message: Messages.encodeConfigApply())
        )

        let deadline = MonotonicClock.nowMs() + timeoutMs
        while true {
            let status = try Messages.decodeConfigStatusRsp(
                try await session.request(path: Messages.pathConfig, message: Messages.encodeConfigGetStatus())
            )
            if status.state == .connected || status.state == .failed { return status }
            guard MonotonicClock.nowMs() <= deadline else { throw FlowError("connection did not settle") }
            try await Task.sleep(nanoseconds: 1_000 * 1_000_000)
        }
    }

    /// Write one DER credential to the device in ordered chunks.
    public static func pushCredential(
        _ session: ProvSession,
        kind: CredKind,
        der: Data,
        chunk: Int = 256
    ) async throws {
        let bytes = [UInt8](der)
        let total = bytes.count
        var off = 0
        while off < total {
            let piece = Data(bytes[off..<min(off + chunk, total)])
            let received = try Messages.decodeCredWriteRsp(
                try await session.request(
                    path: Messages.pathCred,
                    message: Messages.encodeCredWrite(kind: kind, offset: off, total: total, data: piece)
                )
            )
            off += piece.count
            guard received == off else {
                throw FlowError("cred write desync: device has \(received), sent \(off)")
            }
        }
    }

    /// Push device cert + key (+ optional CA), then finalize.
    public static func bootstrapCredentials(
        _ session: ProvSession,
        certDer: Data,
        keyDer: Data,
        caDer: Data? = nil
    ) async throws {
        try await pushCredential(session, kind: .deviceCert, der: certDer)
        try await pushCredential(session, kind: .privateKey, der: keyDer)
        if let caDer {
            try await pushCredential(session, kind: .caCert, der: caDer)
        }
        try Messages.decodeCredFinalizeRsp(
            try await session.request(path: Messages.pathCred, message: Messages.encodeCredFinalize())
        )
    }

    public static func endSession(_ session: ProvSession) async throws {
        try Messages.decodeCtrlRsp(
            try await session.request(path: Messages.pathCtrl, message: Messages.encodeCtrl(.end)),
            op: .end
        )
    }
}
